src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAProjectWithAllFeatures.cs

```diff
@@ -99,14 +81,94 @@ public void It_publishes_the_project_correctly(string targetFramework, string []
+
+            // do a separate build invocation before publish
+            var buildCommand = new BuildCommand(Log, publishCommand.ProjectRootPath);
+            buildCommand.Execute().Should().Pass();
+
+            // modify all project files, which would force recompilation if we were to build during publish
+            foreach (string projectFile in EnumerateFiles(buildCommand, "*.csproj"))
+            {
+                File.AppendAllText(projectFile, " ");
+            }
+
+            // capture modification time of all binaries before publish
+            var modificationTimes = GetLastWriteTimesUtc(buildCommand, "*.exe", "*.dll", "*.resources", "*.pdb");
+
+            // publish (with NoBuild set)
+            publishCommand.Execute("/p:NoBuild=true").Should().Pass();
+            publishCommand.GetOutputDirectory(targetFramework).Should().OnlyHaveFiles(expectedPublishFiles);
+
+            // check that publish did not modify any of the build output
+            foreach (var (file, modificationTime) in modificationTimes)
+            {
+                File.GetLastWriteTimeUtc(file)
+                    .Should().Be(
+                        modificationTime, 
+                        because: $"Publish with NoBuild=true should not overwrite {file}");
+            }
+        }
+
+        private static List<(string, DateTime)> GetLastWriteTimesUtc(MSBuildCommand command, params string[] searchPatterns)
+        {
+            return EnumerateFiles(command, searchPatterns)
+                .Select(file => (file, File.GetLastWriteTimeUtc(file)))
+                .ToList();
+        }
+
+        private static IEnumerable<string> EnumerateFiles(MSBuildCommand command, params string[] searchPatterns)
+        {
+            return searchPatterns.SelectMany(
+                pattern => Directory.EnumerateFiles(
+                    Path.Combine(command.ProjectRootPath, ".."),
+                    pattern,
+                    SearchOption.AllDirectories));
+        }
+
+        private PublishCommand GetPublishCommand(string targetFramework, [CallerMemberName] string callingMethod = null)
+        {
+            TestAsset testAsset = _testAssetsManager
+                .CopyTestAsset("KitchenSink", callingMethod, identifier: targetFramework)
+                .WithSource()
+                .WithProjectChanges((path, project) =>
+                {
+                    if (Path.GetFileName(path).Equals("TestApp.csproj", StringComparison.OrdinalIgnoreCase))
+                    {
+                        var ns = project.Root.Name.Namespace;
+
+                        var targetFrameworkElement = project.Root.Elements(ns + "PropertyGroup").Elements(ns + "TargetFramework").Single();
+                        targetFrameworkElement.SetValue(targetFramework);
+                    }
+                });
+
+            testAsset.Restore(Log, "TestApp");
+            testAsset.Restore(Log, "TestLibrary");
```

Comment:
It shouldn't be necessary to restore TestLibrary separately from TestApp.  (I know this is old code that you just moved)

Reply:
True. I'll fix it.

Reply:
Fixed.